File: backend/app/domain/catalog.py
"""Catalogue domain types: categories, products, reviews, and queries.

Frozen dataclasses per `docs/coding-instructions.md` — internal value
objects that don't need Pydantic validation. Prices are always integer
minor units plus an ISO 4217 currency code (see
`docs/adr/ADR-003-managed-aws-data-tier.md`), never a float, so totals
computed downstream stay exact integer arithmetic.
"""

from __future__ import annotations

from dataclasses import dataclass
from enum import Enum


class SortKey(str, Enum):
    """Fields a product listing can be ordered by."""

    NAME = "name"
    PRICE = "price"


class SortDirection(str, Enum):
    """Direction of a product listing sort."""

    ASC = "asc"
    DESC = "desc"


@dataclass(frozen=True)
class Category:
    """The kind of digital good a Product is. Every Product has exactly one."""

    slug: str
    name: str


@dataclass(frozen=True)
class Review:
    """A shopper's written verdict on a Product."""

    id: str
    author: str
    rating: int
    body: str

    def __post_init__(self) -> None:
        if not 1 <= self.rating <= 5:
            raise ValueError("rating must be between 1 and 5")


@dataclass(frozen=True)
class Product:
    """A digital good offered for sale — an e-book, a licence, or a course."""

    id: str
    name: str
    price_minor: int
    currency: str
    short_description: str
    thumbnail_url: str
    category: Category


@dataclass(frozen=True)
class ProductDetail(Product):
    """A `Product` plus the extra content shown on its detail page."""

    long_description: str
    reviews: tuple[Review, ...]


#: Largest page a caller may request. The HTTP layer should surface a request
#: above this as a 422 rather than silently clamping, so a client paging with a
#: too-large limit learns it is doing so instead of quietly getting short pages.
MAX_PAGE_SIZE = 100


@dataclass(frozen=True)
class ProductQuery:
    """Filter, sort, and page parameters for listing the catalogue."""

    name_contains: str | None = None
    category_slug: str | None = None
    sort: SortKey = SortKey.NAME
    direction: SortDirection = SortDirection.ASC
    limit: int = 20
    offset: int = 0

    def __post_init__(self) -> None:
        if self.limit < 0:
            raise ValueError("limit must be >= 0")
        if self.limit > MAX_PAGE_SIZE:
            # Enforced here so every implementation inherits the ceiling. The
            # catalogue is public and unauthenticated (ADR-002), so without a
            # bound `?limit=1000000000` becomes an unbounded scan of the
            # products table the moment this query reaches Postgres in BE-04 —
            # cheap for the caller, expensive for the database.
            raise ValueError(f"limit must be <= {MAX_PAGE_SIZE}")
        if self.offset < 0:
            raise ValueError("offset must be >= 0")


@dataclass(frozen=True)
class ProductPage:
    """A page of `Product`s together with the total count matching the query."""

    items: tuple[Product, ...]
    total: int

File: backend_js/src/domain/catalog.ts
/**
 * Catalogue domain types: categories, products, reviews, and queries.
 *
 * Plain readonly interfaces plus factory functions — the TypeScript
 * equivalent of the frozen dataclasses in `backend/app/domain/catalog.py`.
 * Prices are always integer minor units plus an ISO 4217 currency code (see
 * `docs/adr/ADR-003-managed-aws-data-tier.md`), never a float, so totals
 * computed downstream stay exact integer arithmetic.
 */

/** Fields a product listing can be ordered by. */
export enum SortKey {
  NAME = 'name',
  PRICE = 'price',
}

/** Direction of a product listing sort. */
export enum SortDirection {
  ASC = 'asc',
  DESC = 'desc',
}

/** The kind of digital good a Product is. Every Product has exactly one. */
export interface Category {
  readonly slug: string
  readonly name: string
}

/** A shopper's written verdict on a Product. */
export interface Review {
  readonly id: string
  readonly author: string
  readonly rating: number
  readonly body: string
}

/**
 * Builds a `Review`, enforcing the rating range the domain promises.
 *
 * The check lives in a factory rather than a constructor because these types
 * are structural interfaces; every place that materialises a Review from
 * storage goes through here, which is what makes the invariant real.
 */
export function makeReview(review: Review): Review {
  if (!(review.rating >= 1 && review.rating <= 5)) {
    throw new RangeError('rating must be between 1 and 5')
  }
  return review
}

/** A digital good offered for sale — an e-book, a licence, or a course. */
export interface Product {
  readonly id: string
  readonly name: string
  readonly priceMinor: number
  readonly currency: string
  readonly shortDescription: string
  readonly thumbnailUrl: string
  readonly category: Category
}

/** A `Product` plus the extra content shown on its detail page. */
export interface ProductDetail extends Product {
  readonly longDescription: string
  readonly reviews: readonly Review[]
}

/**
 * Narrows a `Product` to a `ProductDetail`.
 *
 * Structural types carry no runtime tag, so this stands in for Python's
 * `isinstance(product, ProductDetail)` — used by the cache to record what a
 * payload actually is rather than what the calling method asked for.
 */
export function isProductDetail(product: Product): product is ProductDetail {
  return (
    typeof (product as ProductDetail).longDescription === 'string' &&
    Array.isArray((product as ProductDetail).reviews)
  )
}

/**
 * Largest page a caller may request. The HTTP layer surfaces a request above
 * this as a 422 rather than silently clamping, so a client paging with a
 * too-large limit learns it is doing so instead of quietly getting short
 * pages.
 */
export const MAX_PAGE_SIZE = 100

/** Filter, sort, and page parameters for listing the catalogue. */
export interface ProductQuery {
  readonly nameContains: string | null
  readonly categorySlug: string | null
  readonly sort: SortKey
  readonly direction: SortDirection
  readonly limit: number
  readonly offset: number
}

/**
 * Builds a `ProductQuery`, applying the defaults and the paging invariants.
 *
 * The ceiling is enforced here so every repository implementation inherits
 * it. The catalogue is public and unauthenticated (ADR-002), so without a
 * bound `?limit=1000000000` becomes an unbounded scan of the products table
 * the moment the query reaches Postgres — cheap for the caller, expensive
 * for the database.
 */
export function makeProductQuery(query: Partial<ProductQuery> = {}): ProductQuery {
  const resolved: ProductQuery = {
    nameContains: query.nameContains ?? null,
    categorySlug: query.categorySlug ?? null,
    sort: query.sort ?? SortKey.NAME,
    direction: query.direction ?? SortDirection.ASC,
    limit: query.limit ?? 20,
    offset: query.offset ?? 0,
  }
  if (resolved.limit < 0) {
    throw new RangeError('limit must be >= 0')
  }
  if (resolved.limit > MAX_PAGE_SIZE) {
    throw new RangeError(`limit must be <= ${MAX_PAGE_SIZE}`)
  }
  if (resolved.offset < 0) {
    throw new RangeError('offset must be >= 0')
  }
  return resolved
}

/** A page of `Product`s together with the total count matching the query. */
export interface ProductPage {
  readonly items: readonly Product[]
  readonly total: number
}
